### ISSUE TITLE
Architecture-specific .c files should use target attribute

### ISSUE THREAD

**Issue by Myriachan** [zlib-ng/zlib-ng#336]
When using runtime CPU feature detection, as zlib-ng does, it is best to use `__attribute__((__target__)))` to enable special CPU intrinsics instead of command-line options such as `-msse4`.  The command-line options can cause subtle problems with the One-Definition Rule and lead to crashes on CPUs that do not support those features.

Compiler options such as `-msse4` (GCC/Clang) give compilers carte blanche to use SSE4.2 instructions within functions wherever they please, even if those functions do not use intrinsics.  This becomes a problem with inline functions (likewise template functions in C++).

Consider the following Godbolt example:

https://gcc.godbolt.org/z/G2oNqX

Because `-mssse3` is specified, GCC's vectorizer decides to implement the swaps using the SSSE3 instruction `pshufb`, even though it isn't explicitly using `_mm_shuffle_epi8`.

If `SwapTest` were in a header file that is also #included by a source file not using `-mssse3`, this sometimes is a fatal problem.  The linker would be then free to choose between two different instantiations of `SwapTest`.  The linker might choose the SSSE3 version to be linked in, used by the non-SSSE3 source files, and the program would crash on non-SSSE3 machines.

This can happen with functions declared in system headers, which are not under the control of zlib-ng.

The solution here is to not use the command-line options, and instead use `__attribute__((__target__))` on each function allowed to use processor extensions.

Visual Studio does not understand this attribute.  However, Visual Studio also allows using intrinsics without enabling those CPU features on the command line, so just use nothing in Visual Studio.  Hence something like this:

```
#ifdef _MSC_VER
#define TARGET_ATTRIBUTE_SSE4
#else
#define TARGET_ATTRIBUTE_SSE4 __attribute__((__target__("sse4")))
#endif
...
TARGET_ATTRIBUTE_SSE4
static inline long compare258(const unsigned char *const src0, const unsigned char *const src1) {
```

As a final note, for ARM, Clang's target syntax requires a + sign as the first character, e.g. `"+crypto"` not `"crypto"`, while GCC doesn't.  This requires Clang to be special-cased on ARM.

**Comment by Dead2** [comment]
I agree. This has been way way back on my TODO list actually. Feel free to submit a PR.
Best would probably be to do a compile test during configure/cmake and then enable the feature if setting target attributes is supported by the compiler.
So rather than default it to on, default it to off. Remember there are various compilers out there in regular usage that may or may not support such parameters.

**Comment by Dead2** [comment]
It actually looks like target function attributes can't work. Pragmas is a little better, but still has problems at least on older compilers (GCC 4.x for example).

The thing is that some of the include files need to be compiled with -msse2 for example to enable builtins to be present (otherwise it fails completely or gives you a pure-c fallback builtin instead). And we can't use function attributes for include files.
Pragmas are better since they can be set before include files, but unfortunately (at least older) compilers don't set the expected instruction-set defines that are needed to enable the builtins. This can be worked around by hard-coding the defines, but that quickly becomes a huge mess.

I don't think we can realistically do this.

GCC's recommendation is to separate instruction-level stuff into separate compile units, and that is what we already do. (only functions checking for sse42 is in -msse42 compiled files).
In theory we can get problems with include files containing code (and not just defines or function declarations) that gets compiled with sse42, and if it is also used another place (say in a -msse2 file) the compiler/linker _might_ use the sse42 version it already compiled instead of making a new sse2 variant.
I don't know of any instances of this happening today, I guess we'd have to do qemu test runs with i686 and x86_64 without any optional instruction sets available. That could probably be added to travis, with a chance to detect crashes caused by this potential bug.
System include files should not exhibit this problem.

Our own include files could be a problem if they include functions that are not static (at least I think static avoids the problem).
If we find such problems, one possible solution is to put pragmas around that include file to disable and then re-enable the -msse42 for example.

One of reports I found on this: https://gist.github.com/rygorous/f26f5f60284d9d9246f6

TLDR;
- Target function attributes and target pragmas have problems.
- We currently don't know of any instances of this bug in our code.
- We should add a few travis checks to possibly spot such errors if they occur.
- We can solve the problems another way if we find one.
- Good routines for include-file contents can help avoid the problem.

**Comment by Dead2** [comment]
Closing this now.
